File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int> &prices) {
        int ptr = 0, mn = prices[0], mx = prices[0], n = prices.size();
        int ans = 0;
        while (ptr < n) {
            ptr++;
            if (ptr < n) {
                if (prices[ptr] < mn) {
                    ans = max(mx - mn, ans);
                    mn = mx = prices[ptr];
                } else {
                    mx = max(mx, prices[ptr]);
                }
            }
        }
        ans = max(mx - mn, ans);
        return ans;
    }
};

File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        mn, mx, ans = prices[0], prices[0], 0
        for price in prices:
            if price > mx:
                mx = price
            elif price < mn:
                mn = mx = price
            ans = max(mx - mn, ans)
        return ans